src/adapter/src/session.rs

```diff
@@ -694,18 +693,32 @@ impl<T: TimestampManipulation> Session<T> {
         self.vars.is_superuser()
     }
 
-    /// Returns a channel on which to send external metadata to the session.
-    pub fn retain_external_metadata_transmitter(
+    /// Register a receiver which pushes updates of [`ExternalUserMetadata`].
+    pub fn register_external_metadata_transmitter(
         &mut self,
-    ) -> UnboundedSender<ExternalUserMetadata> {
-        self.external_metadata_tx.clone()
+        rx: tokio::sync::watch::Receiver<ExternalUserMetadata>,
+    ) {
+        let old = self.external_metadata_rx.replace(rx);
+        if old.is_some() {
+            tracing::error!("Double register of external metadata transmitter!");
+        }
     }
 
     /// Drains any external metadata updates and applies the changes from the latest update.
     pub fn apply_external_metadata_updates(&mut self) {
-        while let Ok(metadata) = self.external_metadata_rx.try_recv() {
-            self.vars.set_external_user_metadata(metadata);
+        // If no sender is registered then there isn't anything to do.
+        let Some(rx) = &mut self.external_metadata_rx else {
+            return;
+        };
+
+        // If the value hasn't changed then return.
+        if !rx.has_changed().unwrap_or(false) {
+            return;
         }
+
+        // Update our metadata!
+        let metadata = rx.borrow_and_update().clone();
```

Comment:
```suggestion
        // Update our metadata! Note the short critical section (just a clone), to avoid blocking the sender
        let metadata = rx.borrow_and_update().clone();
```

Reply:
Updated!